Sources/Datadog/DatadogCore/Context/AppStatePublisher.swift

```diff
@@ -0,0 +1,145 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal final class AppStatePublisher: ContextValuePublisher {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    /// The default publisher queue.
+    private static let defaultQueue = DispatchQueue(
+        label: "com.datadoghq.app-state-publisher",
+        target: .global(qos: .utility)
+    )
+
+    /// The initial history value.
+    var initialValue: AppStateHistory
+
+    /// The notification center where this publisher observes following `UIApplication` notifications:
+    /// - `.didBecomeActiveNotification`
+    /// - `.willResignActiveNotification`
+    /// - `.didEnterBackgroundNotification`
+    /// - `.willEnterForegroundNotification`
+    private let notificationCenter: NotificationCenter
+
+    /// The date provider for the Application state snapshot timestamp.
+    private let dateProvider: DateProvider
+
+    /// The queue used to serialise access to the `history` and
+    /// to publish the new history.
+    private let queue: DispatchQueue
+
+    /// The current application state history.
+    ///
+    /// To mutate in the `queue` only.
+    private var history: AppStateHistory
+
+    /// The receiver for publishing the state history.
+    ///
+    /// To mutate in the `queue` only.
+    private var receiver: ContextValueReceiver<AppStateHistory>?
+
+    /// Creates a Application state publisher for publishing application state
+    /// history.
+    ///
+    /// - Parameters:
+    ///   - initialState: The initial application state.
+    ///   - queue: The queue for publishing the history.
+    ///   - dateProvider: The date provider for the Application state snapshot timestamp.
+    ///   - notificationCenter: The notification center where this publisher observes `UIApplication` notifications.
+    init(
+        initialState: AppState,
+        queue: DispatchQueue = AppStatePublisher.defaultQueue,
+        dateProvider: DateProvider = SystemDateProvider(),
+        notificationCenter: NotificationCenter = .default
+    ) {
```

Comment:
nit/ As we have a nice doc string now (🏅), I think it's worth adding in that it needs to be called on the main thread, because it calls `application. applicationState` transitively:
```swift
private static var currentApplicationState: UIApplication.State {
   UIApplication.managedShared?.applicationState ?? .active // fallback to most expected state
}
```
WDYT?

```suggestion
    /// Creates a Application state publisher for publishing application state
    /// history.
    ///
    /// **Note**: It must be called on the main thread.
    ///
    /// - Parameters:
    ///   - initialState: The initial application state.
    ///   - queue: The queue for publishing the history.
    ///   - dateProvider: The date provider for the Application state snapshot timestamp.
    ///   - notificationCenter: The notification center where this publisher observes `UIApplication` notifications.
    init(
        initialState: AppState,
        queue: DispatchQueue = AppStatePublisher.defaultQueue,
        dateProvider: DateProvider = SystemDateProvider(),
        notificationCenter: NotificationCenter = .default
    ) {
```

Reply:
Oh indeed, but this should be documented in the extension below, the one that uses `UIKit`.
I didn't know about threading constraint on `UIApplication/applicationState`. This is a good argument to make the subscription/assignment thread-safe as you mentioned [here](https://github.com/DataDog/dd-sdk-ios/pull/916#discussion_r919788572) so we can dispatch the `ApplicationStatePublisher` init and subscription on the main thread!